File: src/modules/user/user_controller.py
from fastapi import APIRouter

from .dtos import UserCreateDTO, UserUpdateDTO

from .services import CreatorService, IndexerService, FinderService, RemoverService, UpdaterService

class UsersController:
    __creator_service: CreatorService
    __indexer_service: IndexerService
    __finder_service: FinderService
    __remover_service: RemoverService
    __updater_service: UpdaterService

    def __init__(self):
        self.router = APIRouter()
        self._add_routes()
        self.__creator_service = CreatorService()
        self.__indexer_service = IndexerService()
        self.__finder_service = FinderService()
        self.__remover_service = RemoverService()
        self.__updater_service = UpdaterService()

    def _add_routes(self):
        self.router.add_api_route("/", self.create_user, methods=["POST"])
        self.router.add_api_route("/", self.find_users, methods=["GET"])
        self.router.add_api_route("/{id}", self.find_user, methods=["GET"])
        self.router.add_api_route("/{username}", self.update_user, methods=["PATCH"])
        self.router.add_api_route("/{username}", self.delete_user, methods=["DELETE"])

    async def find_users(self):
        return await self.__indexer_service.index()

    async def find_user(self):
        return await self.__finder_service.find()

    async def create_user(self, data: UserCreateDTO):
        return await self.__creator_service.create(data)

    async def update_user(self, data: UserUpdateDTO):
        return await self.__updater_service.update(data)

    async def delete_user(self, username: str):
        return await self.__remover_service.remove()

File: src/modules/user/user.controller.ts
import {
  Controller,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
} from '@nestjs/common';
import { CreateUserDto } from './dto/create-user.dto';
import { UpdateUserDto } from './dto/update-user.dto';
import { CreatorService } from './services';
import { FinderService } from './services/finder.service';
import { FindOneParamsDto } from './dto/find-one-params.dto';
import { IndexerService } from './services/indexer.service';
import { UpdaterService } from './services/updater.service';
import { RemoverService } from './services/remover.service';
import * as bcrypt from 'bcrypt';

@Controller('user')
export class UserController {
  constructor(
    private readonly creatorService: CreatorService,
    private readonly finderService: FinderService,
    private readonly indexerService: IndexerService,
    private readonly updaterService: UpdaterService,
    private readonly removerService: RemoverService,
  ) {}

  @Post()
  public async create(@Body() createUserDto: CreateUserDto) {
    const data = {
      ...createUserDto,
      password: await bcrypt.hash(createUserDto.password, 10),
      cpf: await bcrypt.hash(createUserDto.cpf, 10),
    };

    const createdUser = await this.creatorService.create(data);

    return {
      ...createdUser,
      password: undefined,
      cpf: undefined,
    };
  }

  @Get()
  public async findAll() {
    return this.indexerService.index();
  }

  @Get(':email')
  public async findOne(@Param() dto: FindOneParamsDto) {
    return this.finderService.findByEmail(dto.email);
  }

  @Patch(':email')
  public async update(
    @Param() dto: FindOneParamsDto,
    @Body() updateUserDto: UpdateUserDto,
  ) {
    return this.updaterService.updateByEmail(dto.email, updateUserDto);
  }

  @Delete(':email')
  public async remove(@Param() dto: FindOneParamsDto) {
    return this.removerService.removeByEmail(dto.email);
  }
}
